src/core/search/compressed_list.cc

```diff
@@ -0,0 +1,210 @@
+#include "core/search/compressed_list.h"
+
+#include <array>
+
+#include "base/logging.h"
+
+namespace dfly::search {
+
+using namespace std;
+
+CompressedList::Iterator::Iterator(const CompressedList& list) : stash_{}, diffs_{list.diffs_} {
+  ReadNext();
+}
+
+CompressedList::Iterator::Iterator() : stash_{}, diffs_{nullptr, 0} {
+}
+
+uint32_t CompressedList::Iterator::operator*() const {
+  DCHECK(stash_);
+  return *stash_;
+}
+
+CompressedList::Iterator& CompressedList::Iterator::operator++() {
+  ReadNext();
+  return *this;
+}
+
+bool operator==(const CompressedList::Iterator& l, const CompressedList::Iterator& r) {
+  return l.diffs_.data() == r.diffs_.data() && l.diffs_.size() == r.diffs_.size();
+}
+
+bool operator!=(const CompressedList::Iterator& l, const CompressedList::Iterator& r) {
+  return !(l == r);
+}
+
+void CompressedList::Iterator::ReadNext() {
+  if (diffs_.empty()) {
+    diffs_ = {nullptr, 0};
+    return;
+  }
+
+  uint32_t base = stash_.value_or(0);
+  auto [diff, read] = CompressedList::ReadVarLen(diffs_);
+
+  stash_ = base + diff;
+  diffs_.remove_prefix(read);
+}
+
+CompressedList::Iterator CompressedList::begin() const {
+  return Iterator{*this};
+}
+
+CompressedList::Iterator CompressedList::end() const {
+  return Iterator{};
+}
+
+CompressedList::SortedBackInserter::SortedBackInserter(CompressedList* list)
+    : last_{0}, list_{list} {
+}
+
+CompressedList::SortedBackInserter& CompressedList::SortedBackInserter::operator=(uint32_t value) {
+  DCHECK_LE(last_, value);
+  if (value > last_) {
+    list_->PushBackDiff(value - last_);
+    last_ = value;
+  }
+  return *this;
+}
+
+// Simply encode difference and add to end of diffs array
+void CompressedList::PushBackDiff(uint32_t diff) {
+  array<uint8_t, 16> buf;
+  auto diff_span = WriteVarLen(diff, absl::MakeSpan(buf));
+  diffs_.insert(diffs_.end(), diff_span.begin(), diff_span.end());
+}
+
+// Do a linear scan by encoding all diffs to find value
+CompressedList::EntryLocation CompressedList::LowerBound(uint32_t value) const {
+  uint32_t entry = 0, prev_entry = 0;
+  absl::Span<const uint8_t> diffs_tail{diffs_}, entry_span{};
+
+  while (entry < value && !diffs_tail.empty()) {
+    auto [diff, read] = ReadVarLen(diffs_tail);
+
+    prev_entry = entry;
+    entry += diff;
+
+    entry_span = diffs_tail.subspan(0, read);
+    diffs_tail.remove_prefix(read);
+  }
```

Comment:
Could you use your iterators here instead? They are much cooler :)

Reply:
I'll try. The issue is that I want to keep the iterator simple, so I need to calculate the `entry_span` manually with dirty tricks